### PR TITLE
Fix states

### PR DESCRIPTION
**Pull request by jreidinger**
manually tested

### CODE REVIEW

File: src/NCPkgStatusStrategy.cc [hunk 1]
```diff
@@ -101,6 +101,15 @@ bool NCPkgStatusStrategy::setObjectStatus( ZyppStatus newstatus, ZyppSel slbPtr,
 {
     bool ok = false;
 
+    // workaround to clean previous state properly see bnc#916568 comment#8 especially
+    // basically problem is that if going from status taboo to status noinst sat solver
+    // is set to only remove lock and let solver to install it. What is actually expected
+    // by users is soft lock of package or pattern to not install it unless required
+    // so as workaround cycling thrue states, use different state as previous.
```
Comment (mvidner) [inline src/NCPkgStatusStrategy.cc]: note to self: the status enum is defined in [zypp/ui/Status.h
](https://github.com/openSUSE/libzypp/blob/bb67432eab34163ef3e520183792b83757722219/zypp/ui/Status.h#L34)

**Comment by mvidner** [comment]
I guess understanding the fix is hard, but I don't even understand the problem. Here's a screen 
ncurses-patterns where I can press `-` to get a ` - ` status, but `Space` does nothing.
I got there with `yast -i` on 13.2, Filter, Patterns.

**Comment by jreidinger** [comment]
@mvidner for me space switch state of pattern. I test it as written in bug with software proposal which is part of problem as software proposal readd patterns if it is not already removed by user and not solver. Space change it to --- ( hard lock ) and then to empty space which is not install.

**Comment by mvidner** [comment]
Yes, the important detail is to test SLE12 SP1 installation instead of 13.2 running system. I should know better :-/ Now I can reproduce the problem.

File: src/NCPkgStatusStrategy.cc [hunk 2]
```diff
@@ -101,6 +101,15 @@ bool NCPkgStatusStrategy::setObjectStatus( ZyppStatus newstatus, ZyppSel slbPtr,
...
+    if (newstatus == S_NoInst)
+      slbPtr->setStatus( S_Protected );
```
Comment (mvidner) [inline src/NCPkgStatusStrategy.cc]: So this means that what we see in the installation will change
from `a+` (spc) `---` (spc) ` ` 
to `a+` (spc) `---` (spc) `-i-`?

**Comment by lslezak** [comment]
My question is: does it influence also selecting packages? For me it seems that it will create package locks...

IIUC qt-pkg handles patters differently, see https://github.com/libyui/libyui-qt-pkg/blob/9363013550cbb6b410211498428e8b8d779c1de3/src/YQPkgPatternList.cc#L496

@mvidner?

File: src/NCPkgStatusStrategy.cc (hunk 2 above)
Comment (jreidinger) [inline src/NCPkgStatusStrategy.cc]: no, as you see in the end it set S_NoInst...just previous state is different

**Comment by jreidinger** [comment]
@lslezak so qt do not allow taboo pattern? I do not see that switch.....it should not create lock, only soft lock, that say, that user do not want this package to be installed, which is correct from my POV as user deselect `a+` to ` `

**Comment by lslezak** [comment]
OK, I tested it and qt allows to set taboo to patterns. I'm just asking, I have no idea how it works...

File: src/NCPkgStatusStrategy.cc (hunk 2 above)
Comment (mvidner) [inline src/NCPkgStatusStrategy.cc]: AAH, I missed that this means `setStatus` is called twice.

Comment (jreidinger) [inline src/NCPkgStatusStrategy.cc]: exactly, I just unify previous state before calling setStatus(S_NoInst). I hope it is clear from comment.

**Comment by jreidinger** [comment]
@lslezak @mvidner in the end with @gabi2 help solution is to unify behavior with qt packager that have slightly different status switching

File: src/NCPkgStatusStrategy.cc [hunk 3]
```diff
@@ -277,8 +277,9 @@ bool NCPkgStatusStrategy::toggleStatus( ZyppSel slbPtr,
             }
 	    break;
 	case S_AutoInstall:
-	    //Correct transition is S_Taboo -> #254816
-	    newStatus = S_Taboo;
+	    // this used to be taboo before, but now ZYpp supports
+            // saving weak locks (unselected packages)
```
Comment (lslezak) [inline src/NCPkgStatusStrategy.cc]: NP: the indentation is broken (but maybe it's a GitHub issue...)

**Comment by lslezak** [comment]
LGTM

**Comment by gabi2** [comment]
LGTM (patched package is tested and deselecting pattern works like expected now).

**Pull request merged**